allennlp/common/from_params.py

```diff
@@ -288,3 +294,71 @@ def from_params(cls: Type[T], params: Params, **extras) -> T:
                 kwargs = create_kwargs(cls, params, **extras)
 
             return cls(**kwargs)  # type: ignore
+
+    @classmethod
+    def from_pretrained_params(cls: Type[T], params: Params) -> T:
```

Comment:
My solution means that you can construct objects from third-party libraries that you have no access to, with no subclassing needed.  I'd definitely recommend just looking ahead while parsing parameters instead of changing the class hierarchy.

Reply:
I see, so like right here:

https://github.com/allenai/allennlp/blob/master/allennlp/common/from_params.py#L126

add a check that (basically) 

```
name in params and "_pretrained" in params.get(name, {})
```

and then if that's the case do the from_pretrained logic?

Reply:
I would do it _before_ that check, so you don't require that the annotated type has a `from_params` method, but, basically, yes.

Reply:
i.e., this is another case of the huge `elif` block.

Reply:
right, that's what I meant.

Reply:
Yeah, I think we're agreeing here, and this would work, quite simply.